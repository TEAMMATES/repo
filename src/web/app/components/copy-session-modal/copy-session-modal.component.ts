import { Component, Input } from '@angular/core';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { Course } from '../../../types/api-output';
import { FEEDBACK_SESSION_NAME_MAX_LENGTH } from '../../../types/field-validator';

/**
 * Copy current session modal.
 */
@Component({
  selector: 'tm-copy-session-modal',
  templateUrl: './copy-session-modal.component.html',
  styleUrls: ['./copy-session-modal.component.scss'],
})
export class CopySessionModalComponent {

  // const
  FEEDBACK_SESSION_NAME_MAX_LENGTH: number = FEEDBACK_SESSION_NAME_MAX_LENGTH;

  @Input()
  courseCandidates: Course[] = [];

  @Input()
  sessionToCopyCourseId: string = '';

  newFeedbackSessionName: string = '';
  copyToCourseSet: Set<string> = new Set<string>();

  constructor(public activeModal: NgbActiveModal) {}

  /**
   * Fires the copy event.
   */
  copy(): void {
    if(this.validatenewFeedbackSessionName(this.newFeedbackSessionName)){
      this.activeModal.close({
        newFeedbackSessionName: this.newFeedbackSessionName,
        sessionToCopyCourseId: this.sessionToCopyCourseId,
        copyToCourseList: Array.from(this.copyToCourseSet),
      });
    }
  }

  /**
   * Toggles selection of course to copy to in set.
   */
  select(courseId: string): void {
    if (this.copyToCourseSet.has(courseId)) {
      this.copyToCourseSet.delete(courseId);
    } else {
      this.copyToCourseSet.add(courseId);
    }
  }
  /**
   * validation of newFeedbackSessionName for a course to copy to in set.
   */
  validatenewFeedbackSessionName(newFeedbackSessionName: string): boolean {
    if (newFeedbackSessionName !== null) { // Add a null check
        const regex_pattern = /^(?!\s*$).+/;
        if (regex_pattern.test(newFeedbackSessionName)) {
          return true;
          } else {
          return false;
        }
    } else {
       return true; // If newFeedbackSessionName is null, consider it invalid
    }
}
}
