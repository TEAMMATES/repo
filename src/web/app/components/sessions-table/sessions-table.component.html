<div class="row" *ngIf="sessionsTableRowModels.length > 0; else noSessionMessage">
  <div class="col-12">
    <table class="table table-responsive-lg table-striped table-bordered table-hover margin-bottom-0">
      <thead>
      <tr [ngClass]="{ 'bg-primary text-white': headerColorScheme === SessionsTableHeaderColorScheme.BLUE }">
        <th class="sortable-header fit" (click)="sortSessionsTableRowModels(SortBy.COURSE_ID)" *ngIf="columnsToShow.includes(SessionsTableColumn.COURSE_ID)">
          Course ID
          <span *ngIf="sessionsTableRowModelsSortBy !== SortBy.COURSE_ID" [ngClass]="{ 'text-secondary': headerColorScheme === SessionsTableHeaderColorScheme.WHITE }"><i class="fas fa-sort"></i></span>
          <span *ngIf="sessionsTableRowModelsSortBy === SortBy.COURSE_ID && sessionsTableRowModelsSortOrder === SortOrder.DESC"><i class="fas fa-sort-down"></i></span>
          <span *ngIf="sessionsTableRowModelsSortBy === SortBy.COURSE_ID && sessionsTableRowModelsSortOrder === SortOrder.ASC"><i class="fas fa-sort-up"></i></span>
        </th>
        <th class="sortable-header" (click)="sortSessionsTableRowModels(SortBy.FEEDBACK_SESSION_NAME)">
          Session Name
          <span *ngIf="sessionsTableRowModelsSortBy !== SortBy.FEEDBACK_SESSION_NAME" [ngClass]="{ 'text-secondary': headerColorScheme === SessionsTableHeaderColorScheme.WHITE }"><i class="fas fa-sort"></i></span>
          <span *ngIf="sessionsTableRowModelsSortBy === SortBy.FEEDBACK_SESSION_NAME && sessionsTableRowModelsSortOrder === SortOrder.DESC"><i class="fas fa-sort-down"></i></span>
          <span *ngIf="sessionsTableRowModelsSortBy === SortBy.FEEDBACK_SESSION_NAME && sessionsTableRowModelsSortOrder === SortOrder.ASC"><i class="fas fa-sort-up"></i></span>
        </th>
        <th class="sortable-header" (click)="sortSessionsTableRowModels(SortBy.START_DATE)" *ngIf="columnsToShow.includes(SessionsTableColumn.START_DATE)">
          Start Date
          <span *ngIf="sessionsTableRowModelsSortBy !== SortBy.START_DATE" [ngClass]="{ 'text-secondary': headerColorScheme === SessionsTableHeaderColorScheme.WHITE }"><i class="fas fa-sort"></i></span>
          <span *ngIf="sessionsTableRowModelsSortBy === SortBy.START_DATE && sessionsTableRowModelsSortOrder === SortOrder.DESC"><i class="fas fa-sort-down"></i></span>
          <span *ngIf="sessionsTableRowModelsSortBy === SortBy.START_DATE && sessionsTableRowModelsSortOrder === SortOrder.ASC"><i class="fas fa-sort-up"></i></span>
        </th>
        <th class="sortable-header" (click)="sortSessionsTableRowModels(SortBy.END_DATE)" *ngIf="columnsToShow.includes(SessionsTableColumn.END_DATE)">
          End Date
          <span *ngIf="sessionsTableRowModelsSortBy !== SortBy.END_DATE" [ngClass]="{ 'text-secondary': headerColorScheme === SessionsTableHeaderColorScheme.WHITE }"><i class="fas fa-sort"></i></span>
          <span *ngIf="sessionsTableRowModelsSortBy === SortBy.END_DATE && sessionsTableRowModelsSortOrder === SortOrder.DESC"><i class="fas fa-sort-down"></i></span>
          <span *ngIf="sessionsTableRowModelsSortBy === SortBy.END_DATE && sessionsTableRowModelsSortOrder === SortOrder.ASC"><i class="fas fa-sort-up"></i></span>
        </th>
        <th class="fit">Submissions</th>
        <th class="fit">Responses</th>
        <th class="fit" ngbTooltip="Number of students submitted / Class size" container="body"><span class="tool-tip-decorate">Response Rate</span></th>
        <th class="text-center fit">Action(s)</th>
      </tr>
      </thead>
      <tbody>
      <tr *ngFor="let sessionsTableRowModel of sessionsTableRowModels; let idx = index">
        <td class="fit" *ngIf="columnsToShow.includes(SessionsTableColumn.COURSE_ID)">{{ sessionsTableRowModel.feedbackSession.courseId }}</td>
        <td>{{sessionsTableRowModel.feedbackSession.feedbackSessionName }}</td>
        <td class="fit" *ngIf="columnsToShow.includes(SessionsTableColumn.START_DATE)" container="body"
            [ngbTooltip]="sessionsTableRowModel.feedbackSession.submissionStartTimestamp | formatDateDetail: sessionsTableRowModel.feedbackSession.timeZone">
          <span class="tool-tip-decorate">{{ sessionsTableRowModel.feedbackSession.submissionStartTimestamp | formatDateBrief: sessionsTableRowModel.feedbackSession.timeZone }}</span></td>
        <td class="fit" *ngIf="columnsToShow.includes(SessionsTableColumn.END_DATE)" container="body"
            [ngbTooltip]="sessionsTableRowModel.feedbackSession.submissionEndTimestamp | formatDateDetail: sessionsTableRowModel.feedbackSession.timeZone">
          <span class="tool-tip-decorate">{{ sessionsTableRowModel.feedbackSession.submissionEndTimestamp | formatDateBrief: sessionsTableRowModel.feedbackSession.timeZone }}</span></td>
        <td class="fit" [ngbTooltip]="sessionsTableRowModel.feedbackSession.submissionStatus | submissionStatusTooltip" container="body">
          <span class="tool-tip-decorate">{{ sessionsTableRowModel.feedbackSession.submissionStatus | submissionStatusName }}</span>
        </td>
        <td class="fit" [ngbTooltip]="sessionsTableRowModel.feedbackSession.publishStatus | publishStatusTooltip" container="body">
          <span class="tool-tip-decorate">{{ sessionsTableRowModel.feedbackSession.publishStatus | publishStatusName }}</span>
        </td>
        <td class="fit">
          <a href="#" *ngIf="sessionsTableRowModel.responseRate.length === 0 && !sessionsTableRowModel.isLoadingResponseRate" (click)="loadResponseRateEvent.emit(idx); $event.preventDefault()">Show</a>
          <div *ngIf="sessionsTableRowModel.responseRate.length !== 0">{{ sessionsTableRowModel.responseRate }}</div>
          <tm-ajax-loading *ngIf="sessionsTableRowModel.isLoadingResponseRate"></tm-ajax-loading>
        </td>
        <td class="actions-cell fit">
          <div ngbDropdown class="d-inline-block" ngbTooltip="View the submitted responses for this feedback session">
            <button class="btn btn-primary btn-sm" ngbDropdownToggle>Results</button>
            <div ngbDropdownMenu>
              <button class="dropdown-item clickable" (click)="viewSessionResultEvent.emit(idx)">View Results</button>
              <button class="dropdown-item clickable" ngbTooltip="Make responses no longer visible" placement="left" container="body"
                      *ngIf="![FeedbackSessionSubmissionStatus.NOT_VISIBLE, FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN].includes(sessionsTableRowModel.feedbackSession.submissionStatus)
                        && sessionsTableRowModel.feedbackSession.publishStatus === FeedbackSessionPublishStatus.PUBLISHED; else publishButton"
                      [disabled]="!sessionsTableRowModel.instructorPrivilege.canModifySession" (click)="unpublishSession(idx)">Unpublish Results</button>
              <ng-template #publishButton>
                <button class="dropdown-item clickable" [disabled]="[FeedbackSessionSubmissionStatus.NOT_VISIBLE, FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN].includes(sessionsTableRowModel.feedbackSession.submissionStatus)
                  || sessionsTableRowModel.feedbackSession.publishStatus === FeedbackSessionPublishStatus.PUBLISHED || !sessionsTableRowModel.instructorPrivilege.canModifySession"
                        ngbTooltip="Make session responses available for viewing" placement="left" container="body" (click)="publishSession(idx)">Publish Results</button>
              </ng-template>
              <button class="dropdown-item clickable" *ngIf="sessionsTableRowModel.feedbackSession.publishStatus === FeedbackSessionPublishStatus.PUBLISHED" (click)="remindResultsLinkToStudent(idx)">Resend link to view results</button>
              <button class="dropdown-item clickable">Download Results</button>
            </div>
          </div>
          <button class="btn btn-primary btn-sm" ngbTooltip="Edit feedback session details" (click)="editSessionEvent.emit(idx)"
                  [disabled]="!sessionsTableRowModel.instructorPrivilege.canModifySession">Edit</button>
          <div ngbDropdown class="d-inline-block" ngbTooltip="Actions">
            <button class="action-dropdown btn btn-sm" ngbDropdownToggle><i class="fas fa-ellipsis-v"></i></button>
            <div ngbDropdownMenu>
              <button class="dropdown-item clickable" ngbTooltip="Start submitting feedback"
                      [disabled]="sessionsTableRowModel.feedbackSession.submissionStatus !== FeedbackSessionSubmissionStatus.OPEN
                      || !sessionsTableRowModel.instructorPrivilege.canSubmitSessionInSections" (click)="submitSessionAsInstructorEvent.emit(idx)" >Submit</button>
              <button class="dropdown-item clickable" ngbTooltip="Delete the feedback session" (click)="moveSessionToRecycleBin(idx)"
                      [disabled]="!sessionsTableRowModel.instructorPrivilege.canModifySession">Delete</button>
              <button class="dropdown-item clickable" ngbTooltip="Copy feedback session details" (click)="copySession(idx)">Copy</button>
              <button class="dropdown-item clickable" (click)="sendRemindersToStudents(idx)"
                      [disabled]="sessionsTableRowModel.feedbackSession.submissionStatus !== FeedbackSessionSubmissionStatus.OPEN ||
                          !sessionsTableRowModel.instructorPrivilege.canModifySession">Remind</button>
            </div>
          </div>
        </td>
      </tr>
      </tbody>
    </table>
  </div>
</div>

<ng-template #noSessionMessage>
  <div class="no-session-message">
    There are no feedback sessions in this course.
  </div>
</ng-template>
