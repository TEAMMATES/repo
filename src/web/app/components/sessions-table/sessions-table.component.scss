.no-session-message {
  padding: 10px;
  text-align: center;
}

@media (min-width: 1200px) {
  .actions-cell {
    min-width: 410px;
  }
}

.actions-cell button:not(.dropdown-item) {
  margin-left: var(--btn-margin);
}

.sortable-header {
  cursor: pointer;
}

.clickable {
  cursor: pointer;
}

.margin-bottom-0 {
  margin-bottom: 0;
}

.align-row {
  display: flex;
  flex-direction: row;
}

.loading {
  margin-left: 6px;
  margin-top: 3px;
  width: 12px;
  height: 12px;
  border: 2px solid transparent;
  border-top-color: black;
  border-radius: 50%;
  animation: loading-spinner 1s ease infinite;
}

@keyframes loading-spinner {
  from {
    transform: rotate(0turn);
  }

  to {
    transform: rotate(1turn);
  }
}
