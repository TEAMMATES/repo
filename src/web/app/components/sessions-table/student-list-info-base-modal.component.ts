import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { HttpRequestService } from '../../../services/http-request.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { ErrorMessageOutput } from '../../error-message-output';
import { StudentFeedbackSessionResponseStatus, StudentsFeedbackSessionResponseStatus } from '../../student';
import { SortBy, SortOrder, StudentStatusTableRowModel } from './sessions-table-model';

/**
 * The base modal information for a list of students.
 */
export abstract class StudentListInfoBaseModalComponent {

  // enum
  protected SortBy: typeof SortBy = SortBy;
  protected SortOrder: typeof SortOrder = SortOrder;

  protected studentStatusTableRows: StudentStatusTableRowModel[] = [];
  protected studentsTableRowSortBy: SortBy = SortBy.NONE;
  protected studentsTableRowSortOrder: SortOrder = SortOrder.DESC;

  protected loading: boolean = false;
  protected isAjaxSuccess: boolean = true;

  protected constructor(protected activeModal: NgbActiveModal, protected httpRequestService: HttpRequestService,
                        protected statusMessageService: StatusMessageService) { }

  /**
   * Get the list of students to remind in table form.
   */
  getStudentStatusTableRowModel(paramMap: { [key: string]: string }, model: StudentStatusTableRowModel[]): void {
    this.httpRequestService.get('/session/remind/submission', paramMap)
      .subscribe((studentsFeedbackSessionResponseStatus: StudentsFeedbackSessionResponseStatus) => {
        studentsFeedbackSessionResponseStatus.studentsFeedbackSessionResponseStatus
          .forEach((studentFeedbackSessionResponseStatus: StudentFeedbackSessionResponseStatus) => {

            const studentStatusTableRowModel: StudentStatusTableRowModel = {
              studentFeedbackSessionResponseStatus,
              isChecked: false,
            };

            model.push(studentStatusTableRowModel);
          });
        this.sortStudentsTableRows(SortBy.SUBMIT_STATUS);
      }, (resp: ErrorMessageOutput) => {
        this.statusMessageService.showErrorMessage(resp.error.message);
        this.isAjaxSuccess = false;
      });
    this.loading = false;
  }

  /**
   * Sort the students in according to selection option.
   */
  sortStudentsTableRows(by: SortBy): void {
    this.studentsTableRowSortBy = by;
    // reverse the sort order
    this.studentsTableRowSortOrder =
        this.studentsTableRowSortOrder === SortOrder.DESC ? SortOrder.ASC : SortOrder.DESC;
    this.studentStatusTableRows.sort(this.sortRowsBy(by, this.studentsTableRowSortOrder));
  }

  /**
   * Sorts the rows of students in order.
   */
  sortRowsBy(by: SortBy, order: SortOrder):
      ((a: { studentFeedbackSessionResponseStatus: StudentFeedbackSessionResponseStatus },
        b: { studentFeedbackSessionResponseStatus: StudentFeedbackSessionResponseStatus }) => number) {
    return ((a: { studentFeedbackSessionResponseStatus: StudentFeedbackSessionResponseStatus },
             b: { studentFeedbackSessionResponseStatus: StudentFeedbackSessionResponseStatus }): number => {
      let strA: string;
      let strB: string;
      switch (by) {
        case SortBy.SECTION_NAME:
          strA = a.studentFeedbackSessionResponseStatus.sectionName;
          strB = b.studentFeedbackSessionResponseStatus.sectionName;
          break;
        case SortBy.TEAM_NAME:
          strA = a.studentFeedbackSessionResponseStatus.teamName;
          strB = b.studentFeedbackSessionResponseStatus.teamName;
          break;
        case SortBy.STUDENT_NAME:
          strA = a.studentFeedbackSessionResponseStatus.name;
          strB = b.studentFeedbackSessionResponseStatus.name;
          break;
        case SortBy.STUDENT_EMAIL:
          strA = a.studentFeedbackSessionResponseStatus.email;
          strB = b.studentFeedbackSessionResponseStatus.email;
          break;
        case SortBy.SUBMIT_STATUS:
          strA = a.studentFeedbackSessionResponseStatus.responseStatus.toString();
          strB = b.studentFeedbackSessionResponseStatus.responseStatus.toString();
          break;
        default:
          strA = '';
          strB = '';
      }
      if (order === SortOrder.ASC) {
        return strA.localeCompare(strB);
      }
      if (order === SortOrder.DESC) {
        return strB.localeCompare(strA);
      }
      return 0;
    });
  }

  /**
   * Check all the checkbox of the students.
   */
  checkAllStudents(model: StudentStatusTableRowModel[], isCheck: boolean): void {
    for (const remindStudentRow of model) {
      remindStudentRow.isChecked = isCheck;
    }
  }

  /**
   * Collates a list of selected students with selected checkbox.
   */
  collateStudentsToSend(model: StudentStatusTableRowModel[]): string[] {
    const remindStudentList: string[] = [];
    for (const studentStatusTableRow of model) {
      if (studentStatusTableRow.isChecked) {
        remindStudentList.push(studentStatusTableRow.studentFeedbackSessionResponseStatus.email);
      }
    }
    return remindStudentList;
  }
}
