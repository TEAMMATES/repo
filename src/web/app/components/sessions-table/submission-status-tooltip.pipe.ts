import { Pipe, PipeTransform } from '@angular/core';
import { FeedbackSessionSubmissionStatus } from '../../../types/api-output';

/**
 * Pipe to handle the display of {@code FeedbackSessionSubmissionStatus}.
 */
@Pipe({
  name: 'submissionStatusTooltip',
})
export class SubmissionStatusTooltipPipe implements PipeTransform {

  /**
   * Transforms {@link FeedbackSessionSubmissionStatus} to a tooltip description.
   */
  transform(status: FeedbackSessionSubmissionStatus): string {
    let msg: string = 'The feedback session';

    switch (status) {
      case FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN:
      case FeedbackSessionSubmissionStatus.OPEN:
      case FeedbackSessionSubmissionStatus.GRACE_PERIOD:
      case FeedbackSessionSubmissionStatus.CLOSED:
        msg += ' is visible to students';
        break;
      case FeedbackSessionSubmissionStatus.NOT_VISIBLE:
        msg += ' is not yet visible to students';
        break;
      default:
    }

    switch (status) {
      case FeedbackSessionSubmissionStatus.NOT_VISIBLE:
      case FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN:
        msg += ', and is waiting to open for submissions';
        break;
      case FeedbackSessionSubmissionStatus.OPEN:
        msg += ', and is open for submissions';
        break;
      case FeedbackSessionSubmissionStatus.CLOSED:
        msg += ', and is closed for submissions';
        break;
      default:
    }

    msg += '.';

    return msg;
  }

}
