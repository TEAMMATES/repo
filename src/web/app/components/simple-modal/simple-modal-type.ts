/**
 * Represents the type of  the modal to be displayed.
 */
export enum SimpleModalType {
  /**
   * Warning modal.
   */
  WARNING,
  /**
   * Danger modal.
   */
  DANGER,
  /**
   * Info modal.
   */
  INFO,
  /**
   * Primary modal.
   */
  PRIMARY,
  /**
   * Neutral modal.
   */
  NEUTRAL,
}
