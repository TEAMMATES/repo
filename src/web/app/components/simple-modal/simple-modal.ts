import { TemplateRef } from '@angular/core';

/**
 * Represents a button for the modal.
 */
export interface SimpleModalButton {
  text: string;
  classes?: string;
  action?: Function;
}

/**
 * The standard cancel button for modal.
 */
export const standardCancelButton: (text?: string) => SimpleModalButton = (text?: string) => {
  return {
    text: text || 'Cancel',
    classes: 'btn-light modal-btn-cancel',
  };
};

/**
 * The standard confirm button for modal.
 */
export const standardConfirmButton: (type: string, action: Function, text?: string) => SimpleModalButton =
    (type: string, action: Function, text?: string) => {
      return {
        action,
        text: text || 'Yes',
        classes: `${type} modal-btn-ok`,
      };
    };

/**
 * Parameters for modal.
 */
export interface SimpleModalOptions {
  type: SimpleModalType;
  header: string | TemplateRef<any>;
  content: string | TemplateRef<any>;
  context?: Record<string, any>;
  buttons?: SimpleModalButton[];
  options?: Record<string, string | boolean>;
}

/**
 * Represents the type of the modal to be displayed.
 */
export enum SimpleModalType {
  /**
   * Warning modal.
   */
  WARNING,
  /**
   * Danger modal.
   */
  DANGER,
  /**
   * Info modal.
   */
  INFO,
  /**
   * Neutral modal.
   */
  NEUTRAL,
}
