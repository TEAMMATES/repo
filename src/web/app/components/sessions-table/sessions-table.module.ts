import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { AjaxLoadingModule } from '../ajax-loading/ajax-loading.module';
import { AjaxPreloadModule } from '../ajax-preload/ajax-preload.module';
import { CopySessionModalModule } from '../copy-session-modal/copy-session-modal.module';
import { TeammatesCommonModule } from '../teammates-common/teammates-common.module';
import { FormatDateBriefPipe } from './format-date-brief.pipe';
import { PublishStatusTooltipPipe } from './publish-status-tooltip.pipe';
import {
  ResendResultsLinkToStudentModalComponent,
} from './resend-results-link-to-student-modal/resend-results-link-to-student-modal.component';
import {
  SendRemindersToStudentModalComponent,
} from './send-reminders-to-student-modal/send-reminders-to-student-modal.component';
import { SessionsTableComponent } from './sessions-table.component';
import { StudentListInfoTableComponent } from './student-list-info-table/student-list-info-table.component';
import { SubmissionStatusTooltipPipe } from './submission-status-tooltip.pipe';

/**
 * Module for sessions table.
 */
@NgModule({
  declarations: [
    SubmissionStatusTooltipPipe,
    PublishStatusTooltipPipe,
    SessionsTableComponent,
    ResendResultsLinkToStudentModalComponent,
    SendRemindersToStudentModalComponent,
    FormatDateBriefPipe,
    StudentListInfoTableComponent,
  ],
  imports: [
    CommonModule,
    AjaxLoadingModule,
    AjaxPreloadModule,
    TeammatesCommonModule,
    NgbModule,
    FormsModule,
    CopySessionModalModule,
  ],
  entryComponents: [
    ResendResultsLinkToStudentModalComponent,
    SendRemindersToStudentModalComponent,
  ],
  exports: [
    SessionsTableComponent,
    StudentListInfoTableComponent,
  ],
})
export class SessionsTableModule { }
