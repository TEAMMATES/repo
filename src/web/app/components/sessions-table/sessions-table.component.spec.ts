// Angular imports
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { Pipe, PipeTransform } from '@angular/core';
import {
  ComponentFixture,
  fakeAsync,
  TestBed,
  tick,
  waitForAsync,
} from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
// Third-party imports
import { NgbModal } from '@ng-bootstrap/ng-bootstrap';
// Application imports
import { SessionsTableColumn, SessionsTableRowModel } from './sessions-table-model';
import { SessionsTableComponent } from './sessions-table.component';
import { SessionsTableModule } from './sessions-table.module';
import { SimpleModalService } from '../../../services/simple-modal.service';
// Types and models
import {
  FeedbackSession,
  FeedbackSessionPublishStatus,
  FeedbackSessionSubmissionStatus,
  InstructorPermissionSet,
  ResponseVisibleSetting,
  SessionVisibleSetting,
} from '../../../types/api-output';
import { CopySessionModalComponent } from '../copy-session-modal/copy-session-modal.component';
import { TeammatesRouterModule } from '../teammates-router/teammates-router.module';
// Mock Pipes
@Pipe({ name: 'formatDateBrief' })
class MockFormatDateBriefPipe implements PipeTransform {
  transform(): string {
    return 'Mock Format Date Brief';
  }
}
@Pipe({ name: 'formatDateDetail' })
class MockFormatDateDetailPipe implements PipeTransform {
  transform(): string {
    return 'Mock Format Date Detail';
  }
}
@Pipe({ name: 'publishStatusName' })
class MockPublishStatusNamePipe implements PipeTransform {
  transform(): string {
    return 'Mock Publish Status Name';
  }
}
@Pipe({ name: 'publishStatusTooltip' })
class MockPublishStatusTooltipPipe implements PipeTransform {
  transform(): string {
    return 'Mock Publish Status Tooltip';
  }
}
@Pipe({ name: 'submissionStatusTooltip' })
class MockSubmissionStatusTooltipPipe implements PipeTransform {
  transform(): string {
    return 'Mock Submission Status Tooltip';
  }
}
@Pipe({ name: 'submissionStatusName' })
class MockSubmissionStatusNamePipe implements PipeTransform {
  transform(): string {
    return 'Mock Submission Status Name';
  }
}
describe('SessionsTableComponent', () => {
  let component: SessionsTableComponent;
  let fixture: ComponentFixture<SessionsTableComponent>;
  let ngbModal: NgbModal;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [
        SessionsTableComponent,
        MockFormatDateBriefPipe,
        MockFormatDateDetailPipe,
        MockPublishStatusNamePipe,
        MockPublishStatusTooltipPipe,
        MockSubmissionStatusTooltipPipe,
        MockSubmissionStatusNamePipe,
      ],
      imports: [
        HttpClientTestingModule,
        RouterTestingModule,
        SessionsTableModule,
        TeammatesRouterModule,
      ],
      providers: [
        SimpleModalService,
        { provide: NgbModal, useValue: ngbModal },
      ],
    }).compileComponents();
  }));
  beforeEach(() => {
    fixture = TestBed.createComponent(SessionsTableComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });
  it('should create', () => {
    expect(component).toBeTruthy();
  });
  const feedbackSession1: FeedbackSession = {
    courseId: 'GOT',
    timeZone: 'Asia/Singapore',
    feedbackSessionName: 'Season 8 Review',
    instructions: 'Fill up all',
    submissionStartTimestamp: 1555232400,
    submissionEndTimestamp: 1555332400,
    gracePeriod: 0,
    sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
    responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
    submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
    publishStatus: FeedbackSessionPublishStatus.NOT_PUBLISHED,
    isClosingEmailEnabled: true,
    isPublishedEmailEnabled: true,
    createdAtTimestamp: 1554967204,
    studentDeadlines: {},
    instructorDeadlines: {},
  };
  const feedbackSession2: FeedbackSession = {
    courseId: 'GOT',
    timeZone: 'Asia/Singapore',
    feedbackSessionName: 'Season 7 Review',
    instructions: 'Fill up all',
    submissionStartTimestamp: 1554232400,
    submissionEndTimestamp: 1554332400,
    gracePeriod: 100,
    sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
    responseVisibleSetting: ResponseVisibleSetting.LATER,
    submissionStatus: FeedbackSessionSubmissionStatus.CLOSED,
    publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
    isClosingEmailEnabled: false,
    isPublishedEmailEnabled: false,
    createdAtTimestamp: 1554967204,
    studentDeadlines: {},
    instructorDeadlines: {},
  };
  const instructorCanEverything: InstructorPermissionSet = {
    canModifyCourse: true,
    canModifySession: true,
    canModifyStudent: true,
    canModifyInstructor: true,
    canViewStudentInSections: true,
    canModifySessionCommentsInSections: true,
    canViewSessionInSections: true,
    canSubmitSessionInSections: true,
  };
  const instructorCannotEverything: InstructorPermissionSet = {
    canModifyCourse: false,
    canModifySession: false,
    canModifyStudent: false,
    canModifyInstructor: false,
    canViewStudentInSections: false,
    canModifySessionCommentsInSections: false,
    canViewSessionInSections: false,
    canSubmitSessionInSections: false,
  };
  const sessionTable1: SessionsTableRowModel = {
    feedbackSession: feedbackSession1,
    responseRate: '8 / 9',
    isLoadingResponseRate: false,
    instructorPrivilege: instructorCanEverything,
  };
  const sessionTable2: SessionsTableRowModel = {
    feedbackSession: feedbackSession2,
    responseRate: '',
    isLoadingResponseRate: true,
    instructorPrivilege: instructorCannotEverything,
  };
  it('should snap like in home page with 2 sessions sorted by start date', () => {
    component.columnsToShow = [SessionsTableColumn.START_DATE, SessionsTableColumn.END_DATE];
    component.sessionsTableRowModels = [sessionTable1, sessionTable2];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
  it('should call copySession when triggered', fakeAsync(() => {
    const spy = jest.spyOn(component.copySessionEvent, 'emit');
    const modalRef = {
      result: Promise.resolve({
        newFeedbackSessionName: 'Copied Session',
        targetCourses: ['Course1', 'Course2'],
      }),
      componentInstance: {},
    } as any;
    jest.spyOn(ngbModal, 'open').mockReturnValue(modalRef);
    component.copySession(0);
    tick();
    expect(ngbModal.open).toHaveBeenCalledWith(CopySessionModalComponent);
    expect(spy).toHaveBeenCalledWith({
      newFeedbackSessionName: 'Copied Session',
      targetCourses: ['Course1', 'Course2'],
      sessionToCopyRowIndex: 0,
    });
  }));
});
