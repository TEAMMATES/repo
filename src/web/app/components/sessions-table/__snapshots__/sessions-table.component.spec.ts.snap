// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionsTableComponent should snap with default fields 1`] = `
<tm-sessions-table
  FeedbackSessionPublishStatus={[Function Object]}
  FeedbackSessionSubmissionStatus={[Function Object]}
  SessionsTableColumn={[Function Object]}
  SessionsTableHeaderColorScheme={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  columnsToShow={[Function Array]}
  copySessionEvent={[Function EventEmitter]}
  courseCandidates={[Function Array]}
  editSessionEvent={[Function EventEmitter]}
  headerColorScheme="0"
  loadResponseRateEvent={[Function EventEmitter]}
  modalService={[Function NgbModal]}
  moveSessionToRecycleBinEvent={[Function EventEmitter]}
  publishSessionEvent={[Function EventEmitter]}
  resendResultsLinkToStudentsEvent={[Function EventEmitter]}
  sendRemindersToStudentsEvent={[Function EventEmitter]}
  sessionsTableRowModels={[Function Array]}
  sessionsTableRowModelsSortBy="0"
  sessionsTableRowModelsSortOrder={[Function Number]}
  sortSessionsTableRowModelsEvent={[Function EventEmitter]}
  submitSessionAsInstructorEvent={[Function EventEmitter]}
  unpublishSessionEvent={[Function EventEmitter]}
  viewSessionResultEvent={[Function EventEmitter]}
>
  <div
    class="no-session-message"
  >
     There are no feedback sessions in this course. 
  </div>
</tm-sessions-table>
`;
